pkg/monitor/agent/agent.go

```diff
@@ -19,131 +19,278 @@ import (
-
-	return server, nil
 }
 
-// Agent represents an instance of a monitor agent. It runs a monitor to read
-// events from the BPF perf ring buffer and provides an interface to also pass
-// in non-BPF events.
+// Agent structure for centralizing the responsibilities of the main events
+// reader.
+// There is some racey-ness around perfReaderCancel since it replaces on every
+// perf reader start. In the event that a MonitorListener from a previous
+// generation calls its cleanup after the start of the new perf reader, we
+// might call the new, and incorrect, cancel function. We guard for this by
+// checking the number of listeners during the cleanup call. The perf reader
+// must have at least one MonitorListener (since it started) so no cancel is called.
+// If it doesn't, the cancel is the correct behavior (the older generation
+// cancel must have been called for us to get this far anyway).
 type Agent struct {
-	mutex     lock.Mutex
-	server1_2 net.Listener
-	monitor   *Monitor
-	queue     chan payload.Payload
+	lock.Mutex
+	models.MonitorStatus
+
+	ctx              context.Context
+	perfReaderCancel context.CancelFunc
+	listeners        map[listener.MonitorListener]struct{}
+
+	events        *ebpf.Map
+	monitorEvents *perf.Reader
 }
 
-// NewAgent creates a new monitor agent
-func NewAgent(ctx context.Context, npages int) (a *Agent, err error) {
+// NewAgent starts a new monitor agent instance which distributes monitor events
+// to registered listeners. It spawns a singleton goroutine reading events from
+// the BPF perf ring buffer and provides an interface to pass in non-BPF events.
+// The instance can be stopped by cancelling ctx, which will stop the perf reader
+// go routine and close all registered listeners.
+// Note that the perf buffer reader is started only when listeners are
+// connected.
+func NewAgent(ctx context.Context, nPages int) (a *Agent, err error) {
+	// assert that we can actually connect the monitor
+	path := oldBPF.MapPath(eventsMapName)
+	eventsMap, err := ebpf.LoadPinnedMap(path)
+	if err != nil {
+		return nil, err
+	}
+
 	a = &Agent{
-		queue: make(chan payload.Payload, option.Config.MonitorQueueSize),
+		ctx:              ctx,
+		listeners:        make(map[listener.MonitorListener]struct{}),
+		perfReaderCancel: func() {}, // no-op to avoid doing null checks everywhere
+		events:           eventsMap,
+		MonitorStatus: models.MonitorStatus{
+			Cpus:     int64(eventsMap.ABI().MaxEntries),
+			Npages:   int64(nPages),
+			Pagesize: int64(os.Getpagesize()),
+		},
 	}
 
-	a.server1_2, err = buildServer(defaults.MonitorSockPath1_2)
-	if err != nil {
-		return
+	return a, nil
+}
+
+// SendEvent distributes an event to all monitor listeners
+func (a *Agent) SendEvent(typ int, event interface{}) error {
+	if a == nil {
+		return nil
 	}
 
-	a.monitor, err = NewMonitor(ctx, npages, a.server1_2)
-	if err != nil {
-		return
+	var buf bytes.Buffer
+
+	if a == nil {
+		return fmt.Errorf("monitor agent is not set up")
```

Comment:
This statement will never be reached since just above there's already a check for `a` being `<nil>` (and it returns `nil` if `true`). I think that we probably want to remove the first `a == nil` check.